src/identity/email-password-identity-provider/EmailPasswordIdentityProviderHandler.ts

```diff
@@ -0,0 +1,55 @@
+import type { IncomingMessage, ServerResponse } from 'http';
+import Koa from 'koa';
+import { getLoggerFor } from '../../logging/LogUtil';
+import { HttpHandler } from '../../server/HttpHandler';
+import type { Guarded } from '../../util/GuardedStream';
+
+interface EmailPasswordIdentityProviderHandlerArgs {
+  optional?: string;
+}
+
+export class EmailPasswordIdentityProviderHandler extends HttpHandler {
+  private readonly app: Koa;
+  private readonly logger = getLoggerFor(this);
+
+  public constructor(args: EmailPasswordIdentityProviderHandlerArgs) {
+    super();
+    Object.assign(this, args);
+    this.app = new Koa();
+    this.app.onerror = (err): void => {
+      throw err;
+    };
+    this.app.use(
+      async(ctx, next): Promise<void> => {
+        this.logger.info(ctx.url);
+        if (ctx.url === '/cool') {
+          await new Promise((resolve): void => {
+            setTimeout((): void => {
+              ctx.body = 'cool dude';
+              resolve();
+            }, 2000);
+          });
+        } else {
+          return await next();
+        }
+      },
+    );
+    this.app.use(
+      async(): Promise<void> => {
+        throw new Error('oop error');
+      },
+    );
```

Comment:
Let's avoid making a `Koa` instance for this / using Koa at all.

Instead, make a specific `HttpHandler` for each path. `canHandle` checks whether the path is right, `handle` performs the action for the path.

Reply:
That seems kinda extreme, doesn't it? Some of these handlers are just "serve a login UI" Does that need a whole HttpHandler. I can still do it if you really don't want to have the Koa dependency, but it will take extra work to figure out how to hook up the networking to the oidc provider without Koa.

Reply:
Not extreme; reusable. If the handler is just “serve a login UI”, then we just wire up a StaticFileHandler that can be reused elsewhere.

Let’s avoid Koa (or any framework) in code we write ourselves; if the OIDC lib makes it difficult, we might have no choice for that part.